### ISSUE TITLE
Add HOWTO: Batch Normalization

### ISSUE THREAD

**Issue by marcvanzee** [google/flax#932]
Batch Normalization is more complicated than most layers because of the mutation of moving averages during training.

There is a Discussion on this already (#921), but it may be nice to add a HOWTO for this since we could add the code below as well. The code below is copied from a Colab by @levskaya, which highlights the general state management API you use for any state-computation in a NN.

```
import numpy as np

import jax
from jax import random, lax, numpy as jnp

import flax
from flax import linen as nn
```

We define a trivial conv + BN layer

```
class Foo(nn.Module):
  train: bool
  filters: int

  @nn.compact
  def __call__(self, x):
    x = nn.Conv(self.filters, (1, 1), use_bias=False, dtype=jnp.float32)(x)
    x = nn.BatchNorm(use_running_average=not self.train,
                     momentum=0.9,
                     epsilon=1e-5,
                     dtype=jnp.float32)(x)
    return x

key = random.PRNGKey(0)
x = jnp.ones((5,4,4,3))

# We instantiate the layer then call its init function to get initial variable collections.
foo_vars = Foo(filters=7, train=True).init(key, x)
foo_vars
```

This returns the following:

```
FrozenDict({
    params: {
        Conv_0: {
            kernel: DeviceArray([[[[ 0.50138927,  0.7354811 ,  0.7896391 , -0.63713336,
                             1.081016  , -0.29067358, -0.3780927 ],
                           [ 0.7357814 ,  0.24682549, -0.55378306,  0.16909008,
                             0.85014457,  1.0167135 ,  0.19896305],
                           [ 1.1461202 ,  0.8548834 , -1.0578486 , -0.6013309 ,
                             0.2501557 ,  0.3332178 , -0.36248836]]]], dtype=float32),
        },
        BatchNorm_0: {
            scale: DeviceArray([1., 1., 1., 1., 1., 1., 1.], dtype=float32),
            bias: DeviceArray([0., 0., 0., 0., 0., 0., 0.], dtype=float32),
        },
    },
    batch_stats: {
        BatchNorm_0: {
            mean: DeviceArray([0., 0., 0., 0., 0., 0., 0.], dtype=float32),
            var: DeviceArray([1., 1., 1., 1., 1., 1., 1.], dtype=float32),
        },
    },
})
```

We explicitly say which variable collections are to be mutated by the apply function, those are then returned as auxilliary variables.

```
y1, new_batch_stats = Foo(filters=7, train=True).apply(foo_vars, x, mutable=['batch_stats'])
new_batch_state
```

This returns the following:

```
FrozenDict({
    batch_stats: {
        BatchNorm_0: {
            mean: DeviceArray([ 0.23832898,  0.18371896, -0.08219925, -0.10693741,
                          0.21813159,  0.10592576, -0.05416182], dtype=float32),
            var: DeviceArray([0.9000007 , 0.90000004, 0.9       , 0.90000004, 0.9000001 ,
                         0.9       , 0.9       ], dtype=float32),
        },
    },
})
```

We stitch together params and batch stats collections to evaluate again.
Normally the params would have been updated by a training step using an optimizer.

```
new_foo_vars = {'params': foo_vars['params'], 'batch_stats': new_batch_stats}
y2, even_newer_batch_stats = Foo(filters=7, train=True).apply(new_foo_vars, x, mutable=['batch_stats'])
```

**Comment by cccntu** [comment]
Hey, thanks for the guide. This is useful.

I found something unintuitive:
```
class SimpleModel(nn.Module):
    @nn.compact
    def __call__(self, x):
        x = nn.BatchNorm()(x)
        x = nn.Dense(1)(x)
        return x
model = SimpleModel()

rng = jax.random.PRNGKey(0)
dummy_input = jnp.ones((1, 10))
params = model.init(rng, dummy_input)
```
`params['params']` gives me:
```
FrozenDict({
    BatchNorm_0: {
        scale: DeviceArray([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.], dtype=float32),
        bias: DeviceArray([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32),
    },
    Dense_0: {
        kernel: DeviceArray([[ 0.15454544],
                     [ 0.5659187 ],
                     [ 0.04327311],
                     [ 0.08302793],
                     [-0.00323147],
                     [-0.12254263],
                     [-0.22088017],
                     [-0.2184048 ],
                     [-0.05149708],
                     [-0.42985576]], dtype=float32),
        bias: DeviceArray([0.], dtype=float32),
    },
})
```
and `params.pop('params')` gives me a tuple:
```
(FrozenDict({
     batch_stats: {
         BatchNorm_0: {
             mean: DeviceArray([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32),
             var: DeviceArray([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.], dtype=float32),
         },
     },
 }),
 FrozenDict({
     BatchNorm_0: {
         scale: DeviceArray([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.], dtype=float32),
         bias: DeviceArray([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32),
     },
     Dense_0: {
         kernel: DeviceArray([[ 0.15454544],
                      [ 0.5659187 ],
                      [ 0.04327311],
                      [ 0.08302793],
                      [-0.00323147],
                      [-0.12254263],
                      [-0.22088017],
                      [-0.2184048 ],
                      [-0.05149708],
                      [-0.42985576]], dtype=float32),
         bias: DeviceArray([0.], dtype=float32),
     },
 }))
```
Why are they different?

BTW, I found this usage here:
https://github.com/google/flax/blob/d068512a932da3e05b822790a591bac391aeab36/examples/imagenet/train.py#L69

**Comment by avital** [comment]
Hi @cccntu -- what do you mean by "different"? `pop` returns two parts: the variable collection that you popped, and the remaining collections still all grouped together, e.g.:

```py
variables = model.init(...)
# assume variables['params'] and variables['batch_stats'] are present here
other_variables, params = variables.pop('params')

# here params == variables['params'], and other_variables['batch_stats'] == variables['batch_stats']
```

Maybe the docstring for `FrozenDict.pop` isn't clear enough? But this is working as intended.

(BTW in your example you wrote `params = model.init(rng, dummy_input)` -- it should probably be `variables = model.init(rng, dummy_input)` as parameters are one of possible multiple variable collections)

Does this help? What could we improve in our documentation so that this would be less confusing?

**Comment by cccntu** [comment]
@avital  Thanks for the reply. 
I said different because I expect it to behave like python `dict`. 
But I get it now, since flax/jax is functional, it make sense to return the remaining dictionary. Thanks for the help and the naming suggestion!

I didn't find the doc for FrozenDict from google search and readthedocs search is extremely slow. So I didn't read the doc. 
I just found it by searching Github and I think it's clear enough.

**Comment by avital** [comment]
> I didn't find the doc for FrozenDict from google search and readthedocs search is extremely slow. So I didn't read the doc.
@marcvanzee -- another data point for "slow RTD" ;)

I don't think we even have reference dosc for FrozenDict on RTD, we probably should. I filed an issue: https://github.com/google/flax/issues/969

**Comment by marcvanzee** [comment]
@cccntu could you tell me a bit more about the slowness of readthedocs? Do you think it is slow in general, or only when you search? What search string did you try? I'd like to see if I can reproduce this.

@avital yes I also noticed we don't have `FrozenDict` in RTD, thanks for filing the issue!

**Comment by cccntu** [comment]
@marcvanzee I think it's slow in general from my experiences with different packages on RTD. But right now search seems to take forever. (the 'search...' has been running for several minute now.)
 Here are some string I just tried: 'BatchNorm', 'FrozenDict',  'dense', 'compact' 
(My internet speed is ~100Mbps rn.)

Here is the link produced by using the search box for `dense`
https://flax.readthedocs.io/en/latest/search.html?q=dense&check_keywords=yes&area=default

**Comment by marcvanzee** [comment]
Hi @cccntu, that is surprising. I've heard similar complaints from @apaszke from the JAX team as well, but I haven't experienced it myself (your queries are fast when I try it) and also heard from other Flax users they have no problems with it.

In any case, thanks for reporting it, we will see if this is something that occurs more often, and if so, how we can resolve it.

**Comment by cgarciae** [comment]
Hey! 

I am very interested in the best practices for BatchNorm (or `batch_stats` in general I guess) when used inside a `pmap`. Is the recommendation to use `lax.pmean` to synchronize them? Are there example available?

Thanks!

**Comment by jheek** [comment]
> I am very interested in the best practices for BatchNorm (or batch_stats in general I guess) when used inside a pmap

See the ImageNet example for the canonical example of combining BatchNorm and pmap. There indeed we sync the statistics before evaluation with `pmean`.
